webapp/channels/src/components/suggestion/switch_channel_provider.tsx

```diff
@@ -53,237 +58,267 @@ import {isGuest} from 'mattermost-redux/utils/user_utils';
 import {Preferences} from 'mattermost-redux/constants';
 import {getPreferenceKey} from 'mattermost-redux/utils/preference_utils';
 
-import Provider from './provider';
-import Suggestion from './suggestion.jsx';
+import Provider, {ResultsCallback} from './provider';
+import {SuggestionContainer, SuggestionProps} from './suggestion';
+import {GlobalState} from 'types/store';
 
 const getState = store.getState;
 const searchProfilesMatchingWithTerm = makeSearchProfilesMatchingWithTerm();
-const ThreadsChannel = {
+
+const ThreadsChannel: FakeChannel = {
     id: 'threads',
     name: 'threads',
     display_name: 'Threads',
     type: Constants.THREADS,
     delete_at: 0,
 };
 
-const InsightsChannel = {
+const InsightsChannel: FakeChannel = {
     id: 'insights',
     name: 'activity-and-insights',
     display_name: 'Insights',
     type: Constants.INSIGHTS,
     delete_at: 0,
 };
 
-class SwitchChannelSuggestion extends Suggestion {
-    static get propTypes() {
-        return {
-            ...super.propTypes,
-            channelMember: PropTypes.object,
-            hasDraft: PropTypes.bool,
-            userImageUrl: PropTypes.string,
-            dmChannelTeammate: PropTypes.object,
-            collapsedThreads: PropTypes.bool,
-        };
-    }
+type FakeChannel = Pick<Channel, 'id' | 'name' | 'display_name' | 'delete_at'> & {
+    type: string;
+}
 
-    render() {
-        const {item, isSelection, userImageUrl, status, userItem, collapsedThreads, team, isPartOfOnlyOneTeam} = this.props;
-        const channel = item.channel;
-        const channelIsArchived = channel.delete_at && channel.delete_at !== 0;
-
-        const member = this.props.channelMember;
-        const teammate = this.props.dmChannelTeammate;
-        let badge = null;
-
-        if ((member && member.notify_props) || item.unread_mentions) {
-            let unreadMentions;
-            if (item.unread_mentions) {
-                unreadMentions = item.unread_mentions;
-            } else {
-                unreadMentions = collapsedThreads ? member.mention_count_root : member.mention_count;
-            }
-            if (unreadMentions > 0 && !channelIsArchived) {
-                badge = (
-                    <div className={classNames('suggestion-list_unread-mentions', (isPartOfOnlyOneTeam ? 'position-end' : ''))}>
-                        <span className='badge'>
-                            {unreadMentions}
-                        </span>
-                    </div>
-                );
-            }
-        }
+type FakeDirectChannel = FakeChannel & {
+    userId: string;
+}
 
-        let className = 'suggestion-list__item';
-        if (isSelection) {
-            className += ' suggestion--selected';
-        }
+type ChannelItem = Channel | FakeChannel | FakeDirectChannel;
```

Comment:
This file makes a lot of use of JS's flexibility in a way that makes TS miserable, so I had to define a bunch of different types to make this work while limiting how much I touched the actual logic